core/utils.py

```diff
@@ -1,23 +1,78 @@
 """Utility functions to load data from .mat .yaml and .h5 files."""
 
 import numpy as np
+from jax import numpy as jnp
 import h5py
 import os
 import yaml
 import scipy.io as spio
 import pickle
 from typing import Text
+from pynwb import NWBFile, NWBHDF5IO
+from ndx_pose import PoseEstimationSeries, PoseEstimation
 
+# Constants 
+MM_TO_M = 0.001
 
-def loadmat(filename):
+def daance_to_stac_mjx(data):
     """
-    this function should be called instead of direct spio.loadmat
-    as it cures the problem of not properly recovering python dictionaries
-    from mat files. It calls the function check keys to cure all entries
-    which are still mat-objects
+    Rearrange the data daance .mat mocap ordering
+    to internal order, and scale from millimeters to
+    meters. 
+
+    Args: mocap data arranged in dannce ordering, and
+    stored in mm.
+
+    Out: data rearranged into internal format, scaled to 
+    meters.
+    """
+
+    kp_names = params["KP_NAMES"]
+    # argsort returns the indices that would sort the array
+    stac_keypoint_order = np.argsort(kp_names)
+
+    # Dannce data is stored in mm
+    data = data * MM_TO_M
+    data = jnp.array(data[:, :, stac_keypoint_order])
+    data = jnp.transpose(data, (0, 2, 1))
+    data = jnp.reshape(data, (data.shape[0], -1))
+
+    return data
+
+
+def load_dannce_mat(data_path):
     """
-    data = spio.loadmat(filename, struct_as_record=False, squeeze_me=True)
-    return _check_keys(data)
+    loads in mocap data from .mat file constructed by dannce:
+    (https://github.com/spoonsso/dannce). in particular this means
+    it relies on the data being in millimeters, and that we use the data
+    stored in the "pred" key. 
+    """
+
+    data = spio.loadmat(data_path, struct_as_record=False, squeeze_me=True)
+    data =  _check_keys(data)["pred"][:]
+
+    return daance_to_stac_mjx(data)
+
+
+def load_dannce_nwb(filename):
+    """
+    loads data in from .nwb file. Presumed organized and scaled 
+    equivalent to a dannce .mat file, that has been converted to
+    """
+    data = []
+    with NWBHDF5IO(filename, mode='r', load_namespaces=True) as io:
+        nwbfile = io.read()
+        pose_est = nwbfile.processing['behavior']['PoseEstimation']
+        #print(pose_est.shape)
+
+        node_names = pose_est.nodes[:].tolist()
+
+        for node_name in node_names:
+            data.append(pose_est[node_name].data[:])
+
+        data = np.stack(data, axis=-1)
+
+    return daance_to_stac_mjx(data)
```

Comment:
I think we want to reduce the coupling here. Rather than converting the arrays to the ordering that DANNCE .mat files have when loading, just order it appropriately here.

Tip: I'd use the `node_names` to get them in the ordering specified by the user in the params config.

Reply:
I think I see what you're getting at. Though the intended use case of this function is so that it can be used for both dannce mat files, and nwb files created from them, via the [Colab](https://colab.research.google.com/drive/1CmDhy8XHuHNMBebD15SsuslVdDE4etTx?usp=sharing#scrollTo=mJSWVDRbCkMD) you sent me (linked below). The function not only reorders them, it also scales, transposes, stacks. My intention for a first pass, was to specify in the docs that nwb files would need to the same format as those created from the colab, rather than trying to deal with any number of possible nwb configurations that I imagine could be come up with. It's just meant to be a preliminary effort that will load our own nwb test data. 

I'm still waiting on Charles to lmk if there is a reason he wanted to reorder them. Presuming there isn't, then that still leaves the other reasons I was using this function. Please let me know if there is a reason the other steps aren't valid. 

 

https://colab.research.google.com/drive/1CmDhy8XHuHNMBebD15SsuslVdDE4etTx?usp=sharing#scrollTo=mJSWVDRbCkMD

Reply:
Not sure what you mean. This does the opposite... it takes data stored in dannce format [frames, xyz, keypoint] and converts it to [frames, keypointxyz] used internally. (Ie if the original format is [1000, 3, 10], the final format is [1000, 3*10]. Let me know if I am misunderstanding something.